pkg/pillar/evetpm/tpm.go

```diff
@@ -389,8 +385,225 @@ func readDiskKey(log *base.LogObject) ([]byte, error) {
 	keyBytes, err := tpm2.NVReadEx(rw, TpmDiskKeyHdl,
 		tpm2.HandleOwner, EmptyPassword, 0)
 	if err != nil {
-		log.Errorf("NVReadEx failed: %v", err)
-		return nil, err
+		return nil, fmt.Errorf("NVReadEx failed: %v", err)
 	}
 	return keyBytes, nil
 }
+
+//FetchSealedVaultKey fetches Vault key sealed into TPM2.0
+func FetchSealedVaultKey() ([]byte, error) {
+
+	//gain some knowledge about existing environment
+	sealedKeyPresent := isSealedKeyPresent()
+	legacyKeyPresent := isLegacyKeyPresent()
+
+	if !sealedKeyPresent && !legacyKeyPresent {
+		//Fresh install, go for a new key
+		key, err := GetRandom(vaultKeyLength)
+		if err != nil {
+			return nil, fmt.Errorf("FetchSealedVaultKey: GetRandom failed, %v", err)
+		}
+		err = SealDiskKey(key, DiskKeySealingPCRs)
+		if err != nil {
+			return nil, fmt.Errorf("FetchSealedVaultKey: Sealing failed: %v", err)
+		}
+	}
+
+	if !sealedKeyPresent && legacyKeyPresent {
+		//XXX: we need a migration path for existing installations.
+		//hence re-using the current key here. i.e. if we end up creating
+		//a new random key here, and we fail the upgrade, the fallback
+		//image will not be able to unlock the vault, and we will not be
+		//able to do any upgrade either (since the base image is downloaded
+		//to vault as well).
+		//Hence, we do it in two-steps:
+		//First clone the existing key(not sealed), and make it a sealed key.
+		//In a subsequent release, rotate the sealed key to a new one.
+		//Upgrade path will be to first upgrade to a) first release and then b)
```

Comment:
Do we report the status of being between a and b to the controller?
If not I think we need to add that to report that the vault is looked with a sealed vs. unsealed key, so we and users can easily tell what needs to the updated (or reinstalled).

Reply:
Added a log.Notice, which records the type of the Sealed key (Reused vs New)